### DISCUSSION TITLE
File upload

### DISCUSSION THREAD

**Discussion by timathom** [Jinntec/Fore#245] [Q&A] [answered]
In XForms, there is an `xf:upload` element to render a file upload control. I'm missing an equivalent in Fore. How is uploading handled?

**Comment by timathom** [comment]
Actually, there does seem to be an `fx-upload` control (https://jinntec.github.io/Fore/demo/upload.html)! But this is not included in the documentation or linked from the demo homepage.

**Reply by JoernT** [reply]
I waited for this question for quite some time ;) you hit a hole in our current implementation means it's simply not there yet. It's a bit more effort than just to provide the UI control as upload implies the use of multipart formdata requests to send the files so it has some dependency on fx-submission.

You snippet you found was a starter but it hasn't landed in Fore yet as other priorities always came first and we didn't require it yet in our applications. Now that you requested it i put it high on the list. Please let me know if that's a blocker for you. I'll do the best to push it on the agenda.

**Reply by timathom** [reply]
Thank you! The issue came up because I was interested in designing a simple workflow that would allow a user to upload a file for server-side transformation and then download the result. Having this functionality would be very empowering :)

**Reply by JoernT** [reply, marked as answer]
Well Fore should not hold you back from putting a file input into your page and use it in a usual way.

The interesting point comes when you want to send some additional parameters with the request. That needs binding that to the data. This is already possible with something like this:

```
<fx-control ref="fileupload">
  <label>...</label>
  <input type="file">
</fx-control>
```

If the user chooses a file the filename will end up in the bound node but of course you actually want the data. The actually missing part is multipart formdata support for submission.
